### ISSUE TITLE
T - Tool to automatically to fix fixable flaws

### ISSUE THREAD

**Issue by peterbe** [mdn/yari#680]
If a document has a `broken_links` flaw that, it turns out, is actually a redirect, the general idea is that you edit the `index.html` and replace `<a href="/en-US/docs/Old/Name">` to `<a href="/en-US/docs/Hot/New/Name">`. 

But if the computer can figure out that `'/en-US/docs/Old/Name' => '/en-US/docs/Hot/New/Name'`, why can't it be automated?

This issue is about breaking down such automation. It would be tediuous if you can't do multiple broken links at once so we need a "batch based" solution. 
I think it should be explicit and deliberate. 

note: this may be a unified tool to fix xhrefs and flaws, but still unclear
We can make the first UI a button next to the display of "Broken links" flaws (in the Toolbar) where you can press it to (first dry-run) fix all the references.

**Comment by peterbe** [comment]
I think we should combine this with https://github.com/mdn/yari/issues/695 and expand the scope a bit. #695 talks about easy fixes where the pathname is right but the link accidentally has a full absolute URL in it. And there might be more things that aren't as simple as a redirect. 
For example, 
```html
<a href="/en-uS/doCs/WeB/CsS">
```
will successfully be found in `this.allTitles` but it's spelled incorrectly in its case. 

As discussed [here](https://github.com/mdn/yari/issues/695#issuecomment-638822007) we just need a better name for this.

**Comment by peterbe** [comment]
The reason this is blocked on https://github.com/mdn/yari/issues/623 is because we expect the HTML to only contain `<a href="...">` tags that are actually exactly there in the raw Kuma-HTML. 
What the xrefs macros do is that they turn `{{jsxref('fooBar')}}` into `<a href="/en-US/docs/Foo/Bar">`
So by the time we do `$('a[href]')` and loop over those we'll be encountering links that were never actually typed into the content but instead generated from macros. 

There might be a way to hack on this, until #623 resolves, by (ab)using the raw HTML and do string `.includes` into to see if a hyperlink was raw or came from a KS macro.

**Comment by peterbe** [comment]
Note-to-self; http://localhost:3000/en-US/docs/Web/HTML/Element/img#_flaws is a great example page. It has macros which are redirects, macros which are broken, regular links that a redirects and regular links that are broken.

**Comment by tobinmori** [comment]
Note: likely carryover to sprint Lima.

**Comment by peterbe** [comment]
This issue is turning into an epic (pun intended). 
First of all; of the flaws we currently have code for detecting/logging, all that have hope of being fixed by a computer are now possible to fix by the computer. 

However, the "API" right now is that you have to use the CLI and the `--fix-flaws` option. This is very much similar to using `prettier --write path/to/*.js`. 
That's pretty extreme and not very friendly towards writers. I can imagine some extensions now beyond the CLI option:

1. An env var (e.g. `BUILDER_FIX_FLAWS=true`) that when it runs the builder on every save, it also triggers the fix-flaws functionality. 
2. On the in-browser editor we have that "Auto-save" checkbox option. We can have a "Fix flaws" checkbox option too that is optional. 
3. An action button that appears when you press "Show flaws". But this would require either figuring out, based on the flaws, if anything can be done. Or, you just press it and it can then say "Yes, the content changed"

All of these options are very attractive. They're similar to how I have my VS Code execute `prettier --write` on the file when I ⌘-s so that it saves my edits *plus* a `prettier` cleanup.

**Comment by tobinmori** [comment]
@peterbe how far along are you on this one>? let's talk about this.

**Comment by peterbe** [comment]
> @peterbe how far along are you on this one>? let's talk about this.

I'm a bit stuck right now. Stuck for good ideas. 
There are now merged PRs that fix fixable flaws in macros and in broken links. There are only 2 other flaws: `bad_bcd_queries` and `bad_bcd_links`. Both of which don't have a natural way to automate fixing. 

But there's more we can do "around" this. For example, today the only "API" is that you use the CLI and that might not feel very natural or convenient to writers because they, I roughly assume, are working on a page or a set of pages. Not whole large trees. 

However, one idea is that you enable the CLI flag to always be true so that when the Builder runs via the `server` (remember that involves both the in-browser editor and the watcher) it also fixes fixable flaws. 
But the problem with this is it that it's premature. The CLI was nice because we just had to start somewhere. But for example, if we don't bother with the in-browser editor, then I don't want to waste the effort. Enabling it always in your `.env` might be better but it's quite an important difference and I've love to postpone features like this until we have something more tested beyond just a 1h PVP round. 

So I would say, it's **done in terms of the sprint** and since this the task and [that's the user story](https://github.com/mdn/yari/issues/623) perhaps we close this and keep the user story open for indefinite. But then again, all the juice ideas and discussions have now been happening here in this task issue.
